File: sdks/official/fraiseql-fsharp/tests/FraiseQL.Tests/VectorFieldTests.fs
/// pgvector field authoring: `vector_config` and `vector_distance` (#959).
///
/// The compiler refuses a `Vector` field carrying no `vector_config`, so an SDK that
/// cannot author the config cannot author the type at all. These tests follow the
/// declaration all the way to the exported JSON, because the defect this surface exists
/// to prevent is one that survives authoring and disappears before the compiler sees it
/// — which is what #807 was here, for `requires_scope`.
module FraiseQL.Tests.VectorFieldTests

open System.Text.Json
open Xunit
open FsUnit.Xunit
open FraiseQL

[<GraphQLType(Name = "VectorDocument", SqlSource = "v_document")>]
type VectorDocument() =
    [<GraphQLField(Type = "ID", Nullable = false)>]
    member val Id = "" with get, set

    [<GraphQLField(Type = "Vector",
                   Nullable = false,
                   VectorDimensions = 1536,
                   VectorIndexType = VectorIndex.ivfFlat,
                   VectorDistanceMetric = VectorMetric.l2)>]
    member val Embedding: float array = [||] with get, set

    [<GraphQLField(Type = "BitVector",
                   Nullable = false,
                   VectorDimensions = 768,
                   VectorDistanceMetric = VectorMetric.hamming)>]
    member val Fingerprint = "" with get, set

    [<GraphQLField(Type = "SparseVector",
                   Nullable = true,
                   VectorDimensions = 30000,
                   VectorIndexType = VectorIndex.none)>]
    member val Terms = "" with get, set

    [<GraphQLField(Type = "Vector", Nullable = false, VectorDimensions = 8)>]
    member val Plain: float array = [||] with get, set

    [<GraphQLField(Type = "Float", Nullable = false, VectorDistance = "titleEmbedding")>]
    member val Similarity = 0.0 with get, set

[<GraphQLType(Name = "BothVectorAndDistance", SqlSource = "v_document")>]
type BothVectorAndDistance() =
    [<GraphQLField(Type = "Vector",
                   Nullable = false,
                   VectorDimensions = 8,
                   VectorDistance = "embedding")>]
    member val Embedding: float array = [||] with get, set

[<GraphQLType(Name = "NoDimensions", SqlSource = "v_document")>]
type NoDimensions() =
    [<GraphQLField(Type = "Vector", Nullable = false, VectorDistanceMetric = VectorMetric.l2)>]
    member val Embedding: float array = [||] with get, set

let private field (typeName: string) (fieldName: string) : JsonElement =
    use document = JsonDocument.Parse(SchemaExporter.export ())

    let types =
        document.RootElement.GetProperty("types").EnumerateArray()
        |> Seq.filter (fun t -> t.GetProperty("name").GetString() = typeName)

    let matches =
        types
        |> Seq.collect (fun t -> t.GetProperty("fields").EnumerateArray())
        |> Seq.filter (fun f -> f.GetProperty("name").GetString() = fieldName)
        |> Seq.map (fun f -> f.Clone())
        |> Seq.toList

    match matches with
    | [ f ] -> f
    | _ -> failwithf "field %s.%s is absent from the export" typeName fieldName

let private register () =
    SchemaRegistry.reset ()
    SchemaRegistry.register typeof<VectorDocument>

[<Fact>]
let ``every vector field type carries its own config`` () =
    register ()

    // Every key is asserted, not just the object's presence: index_type and
    // distance_metric both have compiler-side defaults, so a config that lost them
    // would still compile — to hnsw + cosine, chosen by nobody.
    let embedding = (field "VectorDocument" "embedding").GetProperty("vector_config")
    embedding.GetProperty("dimensions").GetInt32() |> should equal 1536
    embedding.GetProperty("index_type").GetString() |> should equal "ivf_flat"
    embedding.GetProperty("distance_metric").GetString() |> should equal "l2"

    (field "VectorDocument" "fingerprint")
        .GetProperty("vector_config")
        .GetProperty("distance_metric")
        .GetString()
    |> should equal "hamming"

    (field "VectorDocument" "terms")
        .GetProperty("vector_config")
        .GetProperty("index_type")
        .GetString()
    |> should equal "none"

[<Fact>]
let ``the index and metric left to the default are written out`` () =
    register ()

    let plain = (field "VectorDocument" "plain").GetProperty("vector_config")
    plain.GetProperty("dimensions").GetInt32() |> should equal 8
    plain.GetProperty("index_type").GetString() |> should equal "hnsw"
    plain.GetProperty("distance_metric").GetString() |> should equal "cosine"

[<Fact>]
let ``a distance reference follows the field name into snake_case`` () =
    // Field names are snake_cased on the way out, so the reference has to be too, or
    // the author points at the name they wrote and the compiler looks for it among
    // names spelled differently.
    register ()

    (field "VectorDocument" "similarity").GetProperty("vector_distance").GetString()
    |> should equal "title_embedding"

[<Fact>]
let ``an ordinary field carries no vector keys`` () =
    register ()

    let id = field "VectorDocument" "id"
    fst (id.TryGetProperty("vector_config")) |> should equal false
    fst (id.TryGetProperty("vector_distance")) |> should equal false

[<Fact>]
let ``a field is an embedding or a distance, not both`` () =
    SchemaRegistry.reset ()

    let thrown =
        Assert.Throws<System.ArgumentException>(fun () ->
            SchemaRegistry.register typeof<BothVectorAndDistance>)

    thrown.Message |> should haveSubstring "not both"

[<Fact>]
let ``a dimension count no column can have is refused`` () =
    SchemaRegistry.reset ()

    let thrown =
        Assert.Throws<System.ArgumentException>(fun () -> SchemaRegistry.register typeof<NoDimensions>)

    thrown.Message |> should haveSubstring "at least 1"

File: sdks/official/fraiseql-csharp/tests/FraiseQL.Tests/VectorFieldTests.cs
using System.Text.Json;
using FraiseQL.Attributes;
using FraiseQL.Export;
using FraiseQL.Models;
using FraiseQL.Registry;
using Xunit;

namespace FraiseQL.Tests;

/// <summary>
/// pgvector field authoring: <c>vector_config</c> and <c>vector_distance</c> (#959).
/// </summary>
/// <remarks>
/// The compiler refuses a <c>Vector</c> field carrying no <c>vector_config</c>, so an SDK
/// that cannot author the config cannot author the type at all. These tests follow the
/// declaration all the way to the exported JSON, because this SDK has already dropped
/// field- and type-level declarations between the registry and the compiler (#849, #807):
/// the registry captured them and the exporter did not carry them across.
/// </remarks>
[Collection(RegistryTestCollection.Name)]
public sealed class VectorFieldTests : IDisposable
{
    public VectorFieldTests() => SchemaRegistry.Instance.Clear();

    public void Dispose() => SchemaRegistry.Instance.Clear();

    [GraphQLType(Name = "Document", SqlSource = "v_document")]
    private sealed class DocumentFixture
    {
        [GraphQLField(Type = "ID")]
        public string Id { get; set; } = string.Empty;

        [GraphQLField(Type = "Vector",
            VectorDimensions = 1536,
            VectorIndexType = VectorIndex.IvfFlat,
            VectorDistanceMetric = VectorMetric.L2)]
        public double[] Embedding { get; set; } = [];

        [GraphQLField(Type = "BitVector",
            VectorDimensions = 768,
            VectorDistanceMetric = VectorMetric.Hamming)]
        public string Fingerprint { get; set; } = string.Empty;

        [GraphQLField(Type = "SparseVector", Nullable = true,
            VectorDimensions = 30000,
            VectorIndexType = VectorIndex.None)]
        public string? Terms { get; set; }

        [GraphQLField(Type = "Vector", VectorDimensions = 8)]
        public double[] Plain { get; set; } = [];

        [GraphQLField(Type = "Float", VectorDistance = "embedding")]
        public double Similarity { get; set; }
    }

    [GraphQLType(Name = "BothVectorAndDistance", SqlSource = "v_document")]
    private sealed class BothVectorAndDistanceFixture
    {
        [GraphQLField(Type = "Vector", VectorDimensions = 8, VectorDistance = "embedding")]
        public double[] Embedding { get; set; } = [];
    }

    [GraphQLType(Name = "NoDimensions", SqlSource = "v_document")]
    private sealed class NoDimensionsFixture
    {
        [GraphQLField(Type = "Vector", VectorDistanceMetric = VectorMetric.L2)]
        public double[] Embedding { get; set; } = [];
    }

    private static JsonElement Field(string typeName, string fieldName)
    {
        using var document = JsonDocument.Parse(SchemaExporter.Export());
        foreach (var type in document.RootElement.GetProperty("types").EnumerateArray())
        {
            if (type.GetProperty("name").GetString() != typeName)
                continue;
            foreach (var field in type.GetProperty("fields").EnumerateArray())
            {
                if (field.GetProperty("name").GetString() == fieldName)
                    return field.Clone();
            }
        }

        throw new Xunit.Sdk.XunitException($"field {typeName}.{fieldName} is absent from the export");
    }

    [Fact]
    public void EveryVectorFieldTypeCarriesItsConfig()
    {
        SchemaRegistry.Instance.Register(typeof(DocumentFixture));

        // Every key is asserted, not just the object's presence: index_type and
        // distance_metric both have compiler-side defaults, so a config that lost them
        // would still compile — to hnsw + cosine, chosen by nobody.
        var embedding = Field("Document", "embedding").GetProperty("vector_config");
        Assert.Equal(1536, embedding.GetProperty("dimensions").GetInt32());
        Assert.Equal("ivf_flat", embedding.GetProperty("index_type").GetString());
        Assert.Equal("l2", embedding.GetProperty("distance_metric").GetString());

        Assert.Equal("hamming", Field("Document", "fingerprint")
            .GetProperty("vector_config").GetProperty("distance_metric").GetString());
        Assert.Equal("none", Field("Document", "terms")
            .GetProperty("vector_config").GetProperty("index_type").GetString());
    }

    [Fact]
    public void TheIndexAndMetricLeftToTheDefaultAreWrittenOut()
    {
        SchemaRegistry.Instance.Register(typeof(DocumentFixture));

        var plain = Field("Document", "plain").GetProperty("vector_config");
        Assert.Equal(8, plain.GetProperty("dimensions").GetInt32());
        Assert.Equal("hnsw", plain.GetProperty("index_type").GetString());
        Assert.Equal("cosine", plain.GetProperty("distance_metric").GetString());
    }

    [Fact]
    public void ADistanceFieldNamesTheVectorItMeasures()
    {
        SchemaRegistry.Instance.Register(typeof(DocumentFixture));

        Assert.Equal("embedding",
            Field("Document", "similarity").GetProperty("vector_distance").GetString());
    }

    [Fact]
    public void AnOrdinaryFieldCarriesNoVectorKeys()
    {
        SchemaRegistry.Instance.Register(typeof(DocumentFixture));

        var id = Field("Document", "id");
        Assert.False(id.TryGetProperty("vector_config", out _));
        Assert.False(id.TryGetProperty("vector_distance", out _));
    }

    [Fact]
    public void AFieldIsAnEmbeddingOrADistanceNotBoth()
    {
        var thrown = Assert.Throws<InvalidOperationException>(
            () => SchemaRegistry.Instance.Register(typeof(BothVectorAndDistanceFixture)));
        Assert.Contains("not both", thrown.Message);
    }

    [Fact]
    public void ADimensionCountNoColumnCanHaveIsRefused()
    {
        var thrown = Assert.Throws<InvalidOperationException>(
            () => SchemaRegistry.Instance.Register(typeof(NoDimensionsFixture)));
        Assert.Contains("at least 1", thrown.Message);
    }
}
